cache/pmdk_cache.cc

```diff
@@ -0,0 +1,438 @@
+    entry->Free();
+  }
+}
+
+void PMDKCacheShard::SetStrictCapacityLimit(bool strict_capacity_limit){
+  // TODO: call transient SetStrictCapacityLimit().
+}
+
+Cache::Handle* PMDKCacheShard::Lookup(const Slice& key, uint32_t hash,
+                                      void* (*pack)(const Slice& slice)) {
+  
+  MutexLock l(&mutex_);
+  // TODO: lookup transient hash table
+
+  // lookup persistent table:
+  po::persistent_ptr<PersistentEntry> e = persistent_hashtable_->Lookup(key, hash);
+  TransientHandle* th = nullptr;
+  if (e != nullptr){
+    th = GetTransientHandle(e, pack);
+    assert(e->InCache());
+    if (!e->HasRefs()){
+      LRU_Remove(e);
+    }
+    th->Ref();
+  }
+  return reinterpret_cast<Cache::Handle*>(th);
+}
+
+bool PMDKCacheShard::Ref(Cache::Handle* h) {
+  TransientHandle* e = reinterpret_cast<TransientHandle*>(h);
+  MutexLock l(&mutex_);
+  // To create another reference - entry must be already externally referenced
+  assert(e->HasRefs());
+  e->Ref();
+  return true;
+}
+
+void PMDKCacheShard::SetHighPriorityPoolRatio(double high_pri_pool_ratio) {
+  // TODO: call transient.
+}
+
+bool PMDKCacheShard::Release(Cache::Handle* handle, bool force_erase) {
+  // TODO
+  return true;
+}
+
+Status PMDKCacheShard::Insert(const Slice& key, uint32_t hash, void* value,
+                             size_t charge,
+                             void (*deleter)(const Slice& key, void* value),
+                             Cache::Handle** handle, Cache::Priority priority,
+                             const Slice& (*unpack)(void* packed),
+                             void* (*pack)(const Slice& value)) {
+  
+  Status s;
+  Status* s_p = &s;
+
+  // TODO: insertion into transient tier, with a dummy deleter that
+  // delays the actual deletion.
+  // note that the transient charge may be underestimated since the
+  // space overhead of (*pack) is not counted. We may argue that it shouldn't be counted...
+
+  autovector<po::persistent_ptr<PersistentEntry>> last_reference_list;
+  // insertion into persistent tier
+  const Slice& unpacked_val = unpack(value);
+  // TODO: better estimation of persistent total charge:
+
+  size_t persistent_charge = sizeof(PersistentEntry) + 
+                              key.size() + 
+                              unpacked_val.size() + 
+                              // rough estimation of hash map entry: cache line size,
+                              // since each hash map node requires its mutex padded
+                              // to cache line.
+                              _SC_LEVEL2_CACHE_LINESIZE;
+
+  po::transaction::run(pop_, [&, unpacked_val, persistent_charge, s_p, deleter, pack] {
+    bool fake_insertion = false;
+    {
+      MutexLock l(&mutex_);
+      EvictFromLRU(persistent_charge, &last_reference_list);
+      // TODO: calculate charge and refuse insert if cache is full
+      if ((usage_ + persistent_charge) > persistent_capacity_ && handle == nullptr){
```

Comment:
Should be `||`?

Reply:
I might be better if I move it (`handle == nullptr`) completely inside. The original logic from `LRUCache` look somewhat like this and I got them mixed up -- sorry.